pkg/operation/botanist/component/vpa/vpa.go

```diff
@@ -0,0 +1,456 @@
+}
+
+func (v *vpa) emptyNetworkPolicy(name string) *networkingv1.NetworkPolicy {
+	return &networkingv1.NetworkPolicy{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: v.namespace}}
+}
+
+func (v *vpa) emptyMutatingWebhookConfiguration() *admissionregistrationv1.MutatingWebhookConfiguration {
+	suffix := "source"
+	if v.values.ClusterType == ClusterTypeShoot {
+		suffix = "target"
+	}
+
+	return &admissionregistrationv1.MutatingWebhookConfiguration{ObjectMeta: metav1.ObjectMeta{Name: "vpa-webhook-config-" + suffix}}
+}
+
+func (v *vpa) rbacNamePrefix() string {
+	prefix := "gardener.cloud:vpa:"
+
+	if v.values.ClusterType == ClusterTypeSeed {
+		return prefix + "source:"
+	}
+
+	return prefix + "target:"
+}
+
+func (v *vpa) serviceAccountNamespace() string {
+	if v.values.ClusterType == ClusterTypeSeed {
+		return v.namespace
+	}
+	return metav1.NamespaceSystem
+}
+
+func getAppLabel(appValue string) map[string]string {
+	return map[string]string{v1beta1constants.LabelApp: appValue}
+}
+
+func getRoleLabel() map[string]string {
+	return map[string]string{v1beta1constants.GardenRole: "vpa"}
+}
+
+func getAllLabels(appValue string) map[string]string {
+	return utils.MergeStringMaps(getAppLabel(appValue), getRoleLabel())
+}
+
+func (v *vpa) getDeploymentLabels(appValue string) map[string]string {
+	if v.values.ClusterType == ClusterTypeSeed {
+		return utils.MergeStringMaps(getAppLabel(appValue), getRoleLabel())
+	}
+
+	return utils.MergeStringMaps(getAppLabel(appValue), map[string]string{
+		v1beta1constants.GardenRole: v1beta1constants.GardenRoleControlPlane,
+	})
+}
+
+type resourceConfig struct {
+	obj      client.Object
+	class    class
+	mutateFn func()
+}
+
+type class uint8
+
+const (
+	runtime class = iota
+	application
+)
+
+type resourceConfigs []resourceConfig
+
+func (r resourceConfigs) allRuntimeObjects() []client.Object {
+	var out []client.Object
+
+	for _, o := range r {
+		if o.class == runtime {
+			out = append(out, o.obj)
+		}
+	}
+
+	return out
+}
+
+func allRuntimeObjects(configsLists ...resourceConfigs) []client.Object {
+	var out []client.Object
+
+	for _, list := range configsLists {
+		out = append(out, list.allRuntimeObjects()...)
+	}
+
+	return out
+}
+
+func mergeResourceConfigs(configsLists ...resourceConfigs) resourceConfigs {
+	var out resourceConfigs
+
+	for _, list := range configsLists {
+		out = append(out, list...)
+	}
+
+	return out
+}
+
+func (v *vpa) injectAPIServerConnectionSpec(deployment *appsv1.Deployment, name string, serviceAccountName *string) {
+	if serviceAccountName != nil {
+		deployment.Spec.Template.Spec.ServiceAccountName = *serviceAccountName
+		deployment.Spec.Template.Spec.Containers[0].Env = append(deployment.Spec.Template.Spec.Containers[0].Env, corev1.EnvVar{
+			Name: "NAMESPACE",
+			ValueFrom: &corev1.EnvVarSource{
+				FieldRef: &corev1.ObjectFieldSelector{
+					FieldPath: "metadata.namespace",
+				},
+			},
+		})
```

Comment:
This injects the `NAMESPACE` variable also in deployments where it wasn't previously present and is likely not needed, e.g. `vpa-recommender`. Perhaps this could be avoided by checking the deployment name?

Reply:
OK, I removed it now from `vpa-recommender`.